README.rst
==========
Wait-a-bit
==========

.. image:: https://img.shields.io/pypi/v/waitabit.svg
        :target: https://pypi.python.org/pypi/waitabit

.. image:: https://img.shields.io/travis/jnevrly/waitabit.svg
        :target: https://travis-ci.org/jnevrly/waitabit

.. image:: https://pyup.io/repos/github/jnevrly/waitabit/shield.svg
     :target: https://pyup.io/repos/github/jnevrly/waitabit/
     :alt: Updates


The simplest waiting line management app - one UI to enter an order number, one
dashboard to display the number (plus a history of previous numbers).

Also, support for a custom HW Keypad (project `Luncheonmeat`_).

The numbers are expected to be assigned separately (manually).

.. figure:: docs/mini_screenshots.png

   UI panel for number input and Dashboard displaying the call queue.


Implementation is a HTML5 and a simple server (Python/Asyncio)
serving both UI screen and the dashboards.

UI screen and dashboards are synchronized over websockets, system allows many
UI screens and many dashboards, but the queue is shared.

UI panel is optimized for a tablet/touchscreen - with swipes and long-press
events.

The system has a screen-saver/session timeout option which would dim the
dashboard and erase the queue after a selected time, to prevent screen burn-in.
Screen-saver can be also enabled manually from the UI. Any new call input
deactivates the screen-saver.

Requirements
------------

* Python 3.5+
* Chrome/Chromium browser (desktop or mobile) - it's probably working on other
  browsers as well, but was not tested.

Installation
------------

Just install the package from GitHub (assuming pip is for Python3)::

    $ sudo pip install git+https://github.com/calcite/waitabit.git

Usage
-----

Running the server
++++++++++++++++++

The server is started from command-line::

    $ waitabit --host 127.0.0.1 --port 8080 --queue_size 6 --session_timeout 3600

The server command has several options apart of mandatory ``host``/``port``:

* ``queue_size`` is the length of the order call history, which is displayed
  on the dashboard.

* ``session_timeout`` is the timeout (in seconds) after which the session
  is closed by clearing the queue and activating the screen-saver.

  Set to 0 to prevent session timeout/screen-saver activation.

* ``keypad_port`` option can be used to specify the `Luncheonmeat` keypad uart
  device path (such as ``/dev/ttyUSB0``).

Full help is also available with ``waitabit --help``.

Using the app
+++++++++++++

The server provides a web-page which links to the UI panel and Dashboard screen.

The dashboard screen doesn't need any user operation, it's just displaying
the queue.

The user input screen is simple, the basic usage is explained
on the image below:

.. image:: docs/input_screen_help.png

Development
-----------

The backend is a standard python package, so you can install it with the usual::

    $ python setup.py develop

The frontend is written in `Vue.js`_ and can be run with the webpack development
server. In the ``waitabit\frontend`` folder, run::

    $ npm install
    $ npm run dev

It will serve the developer view on ``localhost:8080``, but it expects to have
a running backend running on port 8000, so don't forget to have one instance of
waitabit running when starting the npm dev server.

Currently there is no CI set up for this project, so if you wan't to bundle
the distributable JavaScript together with the waitabit python package (as are
all the releases), you need to build the distributable package by running::

    $ npm run build


Licence
-------

* Free software: MIT license

Credits
---------

The ding notification sound is from `Corsica_s`_

The click sound is from `kwahmah_02`_.

The HTML5 part was created with the amazing `Vue.js`_ framework.

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage
.. _`Corsica_S`: https://freesound.org/people/Corsica_S/sounds/91926/
.. _`kwahmah_02`: https://freesound.org/people/kwahmah_02/sounds/256116/
.. _`Vue.js`: https://vuejs.org/
.. _`Luncheonmeat`: https://github.com/calcite/luncheonmeat


HISTORY.rst
=======
History
=======

0.6.0 (2021-03-03)
------------------

* Added support for custom KeyPad device.

0.5.0 (2018-06-12)
------------------

* Added limit on number of digits per call-out number.

0.4.0 (2017-10-10)
------------------

* Added heartbeat to prevent SockJS stalling on Android

3.0 (2017-09-20)
------------------

* Added animation for the new call
* Removed scrollbars

0.2.0 (2017-09-20)
------------------

* Possible to disable session timeout.
* Dashboard panel fullscreen mode.

0.1.0 (2017-09-12)
------------------

* First release



waitabit/cli.py
# -*- coding: utf-8 -*-

"""Console script for waitabit."""

import click
from .waitabit import WaitABit
import asyncio
import logging

@click.command()
@click.version_option()
@click.option('--host', default='0.0.0.0', help="Host IP address. "
                                                "(Default = 0.0.0.0)")
@click.option('--port', default=8000, help="Server port. (Default = 8000)")
@click.option('--queue_size', default=7,
              help="Number of displayed items in the queue. (Default = 7)")
@click.option('--session_timeout', default=3600,
              help="Number of seconds till the session times out"
                   " (queue is cleared and screensaver is enabled)."
                   " Set 0 for no timeout. (Default = 3600)")
@click.option('--heart_beat', default=3,
              help="Heart beat interval [seconds] "
                   "keeping the server connection alive. "
                   "(Default = 3)")
@click.option('--max_digits', default=3,
              help="Maximum number of digits a called-out number can have "
                   "(Default = 3)")
@click.option('--keypad_port', default='off',
              help="serial port to use for connection with custom number entry"
                   "keypad. 'off' value means keypad is not used."
                   "(default = off)")
@click.option('--disable_input_page', default=False, is_flag=True,
              help="Disable number entry via the input panel page.")
def main(host, port, queue_size, session_timeout,
         heart_beat, max_digits, keypad_port, disable_input_page):
    """Wait-a-Bit server."""
    loop = asyncio.get_event_loop()
    srv = WaitABit(queue_size, loop=loop, session_timeout=session_timeout,
                   heart_beat_interval=heart_beat, max_digits=max_digits,
                   keypad_port=keypad_port,
                   disable_input_page=disable_input_page)

    click.echo("Input screen served at http://{0}:{1}/#/input".format(
        host, port))
    click.echo("Info panel served at http://{0}:{1}/#/panel".format(
        host, port))

    srv.run(host=host, port=port)


if __name__ == "__main__":
    main()
